File: ios/Plugin/TcpSocketPlugin.swift
import Foundation
import Capacitor
import SwiftSocket

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(TcpSocketPlugin)
public class TcpSocketPlugin: CAPPlugin {
    var clients: [TCPClient] = []
    
    @objc func connect(_ call: CAPPluginCall) {
        guard let ip = call.getString("ipAddress") else {
            call.reject("Must provide ip address to connect")
            return
        }
        let port = Int32(call.getInt("port", 9100))
        
        let client = TCPClient(address: ip, port: port)
        switch client.connect(timeout: 10) {
          case .success:
            clients.append(client);
            call.resolve(["client": clients.count - 1])
          case .failure(let error):
            call.reject(error.localizedDescription)
        }
    }

    
    @objc func send(_ call: CAPPluginCall) {
        // 获取客户端索引
        let clientIndex = call.getInt("client", -1)
        guard clientIndex >= 0, clientIndex < clients.count else {
            call.reject("No client specified or client index out of range")
            return
        }
        let client = clients[clientIndex]

        // 获取数据类型和数据
        let dataType = call.getString("encoding", "utf8")
        guard let dataString = call.getString("data") else {
            call.reject("No data provided")
            return
        }
        
        print("dataType", dataType)
        print("dataString", dataString)
        
        // 存储转换后的字节数组
        var byteArray = [UInt8]()
        
        // 根据数据类型处理数据
        switch dataType {
        case "utf8":
            byteArray = dataString.utf8.map { $0 }
        case "base64":
            if let decodedData = Data(base64Encoded: dataString) {
                byteArray = [UInt8](decodedData)
            } else {
                call.reject("Invalid Base64 string")
                return
            }
        case "hex":
            let hexString = dataString
            if hexString.count % 2 != 0 {
                call.reject("Invalid hex string length")
                return
            }
            byteArray = stride(from: 0, to: hexString.count, by: 2).compactMap { index in
                let start = hexString.index(hexString.startIndex, offsetBy: index)
                let end = hexString.index(start, offsetBy: 2)
                let byteString = String(hexString[start..<end])
                return UInt8(byteString, radix: 16)
            }
            if byteArray.count * 2 != hexString.count {
                call.reject("Invalid hex string format")
                return
            }
        default:
            call.reject("Unsupported data type")
            return
        }
        
        // 发送数据并处理结果
        switch client.send(data: byteArray) {
        case .success:
            call.resolve()
        case .failure(let error):
            call.reject(error.localizedDescription)
        }
    }
    
    
    @objc func read(_ call: CAPPluginCall) {
        let clientIndex = call.getInt("client", -1)
        if (clientIndex == -1)    {
            call.reject("No client specified")
        }
        let client = clients[clientIndex]
        
        let expectLen = call.getInt("expectLen", 1024)
        let timeout = call.getInt("timeout", 10)
        
        guard let response = client.read(expectLen, timeout: timeout),
                let data = String(bytes: response, encoding: .utf8) else {
            call.resolve(["result": ""])
            return
        }
        
        call.resolve(["result": data])
    }
    
    @objc func disconnect(_ call: CAPPluginCall) {
        let clientIndex = call.getInt("client") ?? -1
        if (clientIndex == -1)  {
            call.reject("No client specified")
        }
        if (clients.indices.contains(clientIndex))   {
            clients[clientIndex].close()
        }
        call.resolve(["client": clientIndex])
    }
}

File: android/src/main/java/com/svend/plugins/tcp/socket/TcpSocketPlugin.java
package com.svend.plugins.tcp.socket;

import android.Manifest;
import android.util.Log;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;

@CapacitorPlugin(name = "TcpSocket", permissions = {@Permission(alias = "network", strings = {Manifest.permission.ACCESS_NETWORK_STATE})})
public class TcpSocketPlugin extends Plugin {

    private Socket socket;
    private DataOutputStream mBufferOut;
    private List<Socket> clients = new ArrayList<>();

    @PluginMethod
    public void connect(PluginCall call) {
        String ipAddress = call.getString("ipAddress");

        if (ipAddress == null || ipAddress.isEmpty()) {
            call.reject("Must provide ip address to connect");
            return;
        }
        Integer port = call.getInt("port", 9100);

        try {
            if (socket != null && socket.isConnected()) {
                socket.close();
            }
            socket = new Socket(ipAddress, port);
            clients.add(socket);
        } catch (IOException e) {
            Log.d("Connection failed", e.getMessage());
            call.reject(e.getMessage());
            return;
        }

        JSObject ret = new JSObject();
        ret.put("client", clients.size() - 1);
        call.resolve(ret);
    }

    @PluginMethod
    public void send(final PluginCall call) {
        final Integer client = call.getInt("client", -1);
        final String msg = call.getString("data", "");
        final String dataType = call.getString("encoding", "utf8"); // 默认使用 utf8

        if (client == -1) {
            call.reject("No client specified");
            return;
        }

        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                try {
                    final Socket socket = clients.get(client);
                    mBufferOut = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream()));
                    byte[] byteArray;

                    switch (dataType) {
                        case "utf8":
                            byteArray = msg.getBytes();
                            break;
                        case "hex":
                            byteArray = hexStringToByteArray(msg);
                            break;
                        default:
                            call.reject("Unsupported data type");
                            return;
                    }

                    if (mBufferOut != null) {
                        mBufferOut.write(byteArray);
                        mBufferOut.flush();
                    }

                    call.resolve();

                } catch (IOException e) {
                    call.reject(e.getMessage());
                }
            }
        };

        Socket socket = clients.get(client);
        if (socket == null || !socket.isConnected()) {
            try {
                if (socket != null) {
                    socket.close();
                }
            } catch (IOException e) {
                call.reject("Generic error");
            }
            call.reject("Socket not connected");
            return;
        }
        Thread thread = new Thread(runnable);
        thread.start();
    }

    private byte[] hexStringToByteArray(String s) {
        int len = s.length();
        if (len % 2 != 0) {
            s = "0" + s; // 如果长度是奇数，前面补0
            len++;
        }
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4) + Character.digit(s.charAt(i + 1), 16));
        }
        return data;
    }

    @PluginMethod
    public void read(final PluginCall call) {
        final Integer client = call.getInt("client", -1);
        final Integer length = call.getInt("expectLen", 1024);

        if (client == -1 || length == -1) {
            call.reject("Client or length not specified");
            return;
        }

        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                try {
                    final Socket socket = clients.get(client);
                    DataInputStream mBufferIn = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
                    byte[] bytes = new byte[length];
                    int read = mBufferIn.read(bytes, 0, length);

                    JSObject ret = new JSObject();
                    ret.put("result", new String(bytes, 0, read));
                    call.resolve(ret);
                } catch (IOException e) {
                    call.reject(e.getMessage());
                }
            }
        };

        Socket socket = clients.get(client);
        if (!socket.isConnected()) {
            try {
                socket.close();
            } catch (IOException e) {
                call.reject("Generic error");
            }
            call.reject("Socket not connected");
            return;
        }
        Thread thread = new Thread(runnable);
        thread.start();
    }

    @PluginMethod
    public void disconnect(PluginCall call) {
        final Integer client = call.getInt("client", -1);
        if (client == -1) {
            call.reject("No client specified");
            return;
        }
        if (clients.isEmpty()) {
            call.reject("Socket not connected");
            return;
        }
        final Socket socket = clients.get(client);
        try {
            if (!socket.isConnected()) {
                socket.close();
                call.reject("Socket not connected");
            }
            socket.close();
        } catch (IOException e) {
            call.reject(e.getMessage());
        }

        JSObject ret = new JSObject();
        ret.put("client", client);
        call.resolve(ret);
    }
}
